netx.go

```diff
@@ -5,27 +5,75 @@ package netx
 import (
 	"context"
 	"net"
+	"strings"
 	"sync/atomic"
 	"time"
 )
 
 var (
-	dial           atomic.Value
-	dialUDP        atomic.Value
-	listenUDP      atomic.Value
-	resolveTCPAddr atomic.Value
-	resolveUDPAddr atomic.Value
-
+	dial               atomic.Value
+	dialUDP            atomic.Value
+	listenUDP          atomic.Value
+	resolveTCPAddr     atomic.Value
+	resolveUDPAddr     atomic.Value
+	NAT64Prefix        atomic.Value
 	defaultDialTimeout = 1 * time.Minute
 )
 
 func init() {
 	Reset()
 }
 
+type NAT64PrefixHolder struct {
+	expiration time.Time
+	prefix     []byte
+}
+
+func getNAT64Prefix() []byte {
+	if holder, ok := NAT64Prefix.Load().(*NAT64PrefixHolder); holder != nil && ok {
+		if time.Now().Before(holder.expiration) {
+			return holder.prefix
+		}
+	}
+	ips, err := net.LookupIP("ipv4only.arpa")
+	if err == nil {
+		for _, ip := range ips {
+			if ip.To4() == nil {
+				prefix := ip[:12]
+				NAT64Prefix.Store(&NAT64PrefixHolder{prefix: prefix, expiration: time.Now().Add(time.Hour * 1)})
+				return prefix
+			}
+		}
+	}
+	return nil
+}
+
+func isNetworkUnreachable(err error) bool {
+	return err != nil && strings.Contains(err.Error(), "network is unreachable")
```

Comment:
This seems fine, but just a note that we should test this is what is triggered on windows, android etc if you haven't already since there tends to be a lack of consistency to the shape of network errors across platforms.

Reply:
@forkner  i've confirmed in Go stdlib source code that the same string is used on all platforms

Reply:
all seems fine in theory, just needs verification that it actually triggers this expected error in practice once it's built on each platform.  My guess is also that this works / is sufficient.  Varied os network stacks underneath mean sometimes it bubbles up some other error in the same circumstances although there is a (mostly) common set of errors it might generate.

Reply:
@forkner it would be a b**ch to test, since we only have one Linux NAT64 machine set up.
Maybe @benjojo can help us with setting up a Windows and OSX boxes with NAT64?

We should also test this on Android/iOS somehow...

Reply:
I run t-mobile LTE on my main Android phone. I've never had trouble connecting to our proxies. What should I test for?

Reply:
Ah, the NAT64 stuff T-Mobile LTE does may be iPhone only then, I'll dig around some more